### PR TITLE
Added lint tests for notification sounds

### PR DESCRIPTION
**Pull request by TamasKiss**
Now running the tests will include checking every notification on actors.

### CODE REVIEW

File: OpenRA.Mods.Common/Lint/CheckNotificationReferences.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+								continue;
+
+							CheckVoices(actorInfo.Value, emitError, rules, notification);
+						}
+					}
+				}
+			}
+		}
+
+		void CheckVoices(ActorInfo actorInfo, Action<string> emitError, Ruleset rules, string notification)
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Lint/CheckNotificationReferences.cs]: Please name this `CheckNotifications` instead.

**Comment by GraionDilach** [comment]
Something is very suspicious to me about the notification check itself and Travis blew up the lint tests.

File: OpenRA.Mods.Common/Lint/CheckNotificationReferences.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		{
+			var soundInfo = rules.Notifications[notification.ToLowerInvariant()];
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckNotificationReferences.cs]: This needs to first check whether `notification` exists in `rules.Notifications`!

File: OpenRA.Mods.Common/Lint/CheckNotifications.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+								continue;
+
+							CheckVoices(actorInfo.Value, emitError, rules, notification);
+						}
+					}
+				}
+			}
+		}
+
+		void CheckVoices(ActorInfo actorInfo, Action<string> emitError, Ruleset rules, string notification)
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Lint/CheckNotifications.cs]: Again, please rename this to CheckNotifications. It's not voices per se what you're checking for.

File: OpenRA.Mods.Common/Lint/CheckNotifications.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+		{
+			if (!rules.Notifications.ContainsKey(notification))
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Lint/CheckNotifications.cs]: Don't we want a lint error in this case as well? Asin "Actor's using undefined Notification string X!"?

**Comment by GraionDilach** [comment]
Also, if you're updating a PR, please leave a comment on doing so, because GitHub does not notify people on new pushes, only on comments.

File: OpenRA.Mods.Common/Traits/Crates/CrateAction.cs [hunk 5]
```diff
@@ -27,6 +27,7 @@ public class CrateActionInfo : ITraitInfo
 		[PaletteReference] public readonly string Palette = "effect";
 
 		[Desc("Audio clip to play when the crate is collected.")]
+		[NotificationReferenceAttribute]
```
Comment (MustaphaTR) [inline OpenRA.Mods.Common/Traits/Crates/CrateAction.cs]: This is not actual Notification reference, but direct sound filename. Bot sure why it is caled Notification.

**Comment by MustaphaTR** [comment]
Travis/make test gives errors for evwry map that has custom rules + 1 for main mod. So actual number of errors indeed wouldn't be that high.

**Comment by TamasKiss** [comment]
CrateAction no longer contains the NotificationReferenceAttribute, as @MustaphaTR suggested. Thank you for the review.

I tried to poke around a bit since yesterday to see what is the root of the many errors, but could not find it. I will try some more tomorrow.

**Comment by TamasKiss** [comment]
Finally got back to it, is `(!rules.Notifications.ContainsKey(notification))` really a cause for throwing an error? Imo a null there is normal.

**Comment by pchote** [comment]
A null (i.e. no notification) should be allowed, but a reference to a notification type that isn't defined definitely shouldn't.

**Comment by TamasKiss** [comment]
I updated the commit, not sure if I understood you right. 

btw the notifications that are causing these errors are mostly:
- UnitLost/Captured
- BuildingLost/Captured
But there are a few other offenders

**Comment by pchote** [comment]
I see that a bunch of notifications that aren't tagged with the attribute yet.  I suggest using "Find References" `Game.Sound.PlayNotification` to list all the places that call it, and then from there find all the info fields that are passed in.

There's also an unintuitive style nit: the compiler contains some magic that supports omitting the `Attribute` part of the class name when it is used inside `[]`s, and that is the preferred c# code style.   So all the `[NotificationReferenceAttribute]`s should change to `[NotificationReference]` (but not the class name!).

**Comment by pchote** [comment]
Any progress here @TamasKiss?

**Comment by TamasKiss** [comment]
Sorry, but no change yet! I can do it in the afternoon though, it skipped my mind. 

Thanks for the reminder!

**Comment by TamasKiss** [comment]
I added the extra notifications and renamed the usages. 

I also modified the namespace of RepairsUnits to mirror its directory.

Thanks for the help!

**Comment by GraionDilach** [comment]
You need to rebase your branch.

**Comment by TamasKiss** [comment]
It should be ready now

**Comment by GraionDilach** [comment]
Think you mixed up branches.

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs [hunk 6]
```diff
@@ -94,6 +86,22 @@ public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRendere
...
 			};
 
+			editorLayer.OnSpawnRemoved = () =>
+			{
+				if (editorLayer.Players.Players.Values.Any(p => p.Name.Equals(selectedOwner.Name)))
+					return;
+
+				var orderedOwners = editorLayer.Players.Players.Values.ToList();
+				orderedOwners.Add(selectedOwner);
+				orderedOwners.Sort((x, y) => String.Compare(x.Name, y.Name, StringComparison.Ordinal));
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: > [SA1121] Use the built-in type alias 'string' rather than String or System.String

**Comment by pchote** [comment]
I'll take this over once #15357 has been merged.

**Comment by pchote** [comment]
Thanks for your help with setting up the base for this @TamasKiss!

I ended up reworking this a fair bit and didn't want to risk breaking your branches (you have other PR code in here) so I filed #15646 as a new PR instead of pushing it over this one.

Lets move the testing and discussion over to the new ticket.

**Pull request closed without merging**